### PR TITLE
Implement Batch, Layer, Instance normalization

### PR DESCRIPTION
**Pull request by vinayhpandya**
Added normalizations to improve stability of models during training
- Added batch Normalization to normalize features across batch
- Added layer normalization to normalize features across layers
- Added Instance normalization to normalize individual node features in a graph

## Proposed changes

Added normalizations to improve stability during training

## Checklist

Put an `x` in the boxes that apply.
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] I have updated the necessary documentation

### CODE REVIEW

File: pyproject.toml [hunk 1]
```diff
@@ -32,6 +32,7 @@ test = [
   "pytest==7.4.4",
   "scipy==1.12.0",
   "networkx==3.2.1",
+  "torch==2.2.0",
```
Comment (francescofarina) [inline pyproject.toml:35]: You need `torch_geometric` here

**Review by francescofarina — CHANGES_REQUESTED** [review]
Can you remove the types in the docstrings' args?

What do we think about naming these new layer using the `Normalization` keyword instead of `Norm`? Like `BatchNormalization`. I always felt that using norm is prone to mislead users.

File: mlx_graphs/nn/norm/batch_norm.py [hunk 2]
```diff
@@ -0,0 +1,77 @@
...
+    The mean and standard-deviation are calculated per-dimension over all nodes
+    inside the mini-batch.
+
+    Args:
+        in_channels : Size of each input sample.
+        eps (float, optional): A value added to the denominator for numerical
+            stability. (default: :obj:`1e-5`)
+        momentum : The value used for the running mean and
+            running variance computation. (default: :obj:`0.1`)
+        affine (bool, optional): If set to :obj:`True`, this module has
```
Comment (francescofarina) [inline mlx_graphs/nn/norm/batch_norm.py:27]: rm type

File: mlx_graphs/nn/norm/layer_norm.py [hunk 3]
```diff
@@ -0,0 +1,108 @@
...
+        \\mathbf{x}^{\\prime}_i = \\frac{\\mathbf{x} -
+        \\textrm{E}[\\mathbf{x}]}{\\sqrt{\\textrm{Var}[\\mathbf{x}] + \\epsilon}}
+        \\odot \\gamma + \\beta
+
+    The mean and standard-deviation are calculated across all nodes and all
+    node channels separately for each object in a mini-batch.
+
+    Args:
+        in_channels : Size of each input sample.
+        eps (float, optional): A value added to the denominator for numerical
```
Comment (francescofarina) [inline mlx_graphs/nn/norm/layer_norm.py:25]: rm type

File: mlx_graphs/nn/norm/batch_norm.py [hunk 4]
```diff
@@ -0,0 +1,77 @@
...
+            with only a single element will work as during in evaluation.
+            That is the running mean and variance will be used.
+            Requires :obj:`track_running_stats=True`. (default: :obj:`False`)
+    """
+
+    def __init__(
+        self,
+        num_features: int,
+        eps: float = 1e-5,
+        momentum: Optional[float] = 0.1,
```
Comment (francescofarina) [inline mlx_graphs/nn/norm/batch_norm.py:45]: A default value is provided, so this is not `Optional`

File: mlx_graphs/nn/norm/layer_norm.py [hunk 5]
```diff
@@ -0,0 +1,108 @@
...
+            normalization (:obj:`"graph"` or :obj:`"node"`). If :obj:`"graph"`
+            is used, each graph will be considered as an element to be
+            normalized. If `"node"` is used, each node will be considered as
+            an element to be normalized. (default: :obj:`"graph"`)
+    """
+
+    def __init__(
+        self,
+        num_features: int,
+        eps: Optional[float] = 1e-5,
```
Comment (francescofarina) [inline mlx_graphs/nn/norm/layer_norm.py:40]: A default value is provided, so this is not `Optional`

**Review by francescofarina — CHANGES_REQUESTED** [review]
I left a few extra comments.

It looks great overall - @TristanBilot what do you think about the naming of the normalization layers

File: mlx_graphs/nn/norm/instance_norm.py [hunk 6]
```diff
@@ -0,0 +1,129 @@
...
+            return self.instance_norm(mx.expand_dims(node_features, axis=0)).squeeze(
+                axis=0
+            )
+
+        if batch_size is None:
+            batch_size = batch.max().item() + 1
+
+        mean = var = unbiased_var = node_features
+        if self.training or not self.track_running_stats:
+            print("inside training")
```
Comment (TristanBilot) [inline mlx_graphs/nn/norm/instance_norm.py:80]: can you remove print please

File: mlx_graphs/nn/norm/layer_norm.py [hunk 7]
```diff
@@ -0,0 +1,108 @@
...
+            normalized. If `"node"` is used, each node will be considered as
+            an element to be normalized. (default: :obj:`"graph"`)
+    """
+
+    def __init__(
+        self,
+        num_features: int,
+        eps: float = 1e-5,
+        affine: bool = True,
+        mode: str = "graph",
```
Comment (TristanBilot) [inline mlx_graphs/nn/norm/layer_norm.py:42]: would be better to have `Literal["graph", "node"] = "graph"` here

File: tests/nn/norm/test_batch_norm.py [hunk 8]
```diff
@@ -0,0 +1,38 @@
...
+    batch_norm.train()
+    node_features = mx.random.uniform(0, 1, [6, 8])
+    print(node_features.shape)
+    normalized = batch_norm(node_features)
+
+    assert normalized.shape == (6, 8)
+    mean = normalized.mean(axis=0)
+    normalized_std = mx.sqrt(normalized.var(axis=0))
+    assert mx.allclose(mean, mx.zeros(8), atol=1e-3)
+    print(normalized_std)
```
Comment (TristanBilot) [inline tests/nn/norm/test_batch_norm.py:20]: rm

File: tests/nn/norm/test_batch_norm.py [hunk 9]
```diff
@@ -0,0 +1,38 @@
...
+from mlx_graphs.nn import BatchNormalization
+
+mx.random.seed(42)
+
+
+def test_batch_norm():
+    batch_norm = BatchNormalization(num_features=8)
+    batch_norm.train()
+    node_features = mx.random.uniform(0, 1, [6, 8])
+    print(node_features.shape)
```
Comment (TristanBilot) [inline tests/nn/norm/test_batch_norm.py:13]: rm

**Review by TristanBilot — CHANGES_REQUESTED** [review]
Very nice PR, just added few reviews!

**Review by francescofarina — APPROVED** [review]
